src/include/storage/copier/compression.h

```diff
@@ -0,0 +1,313 @@
+        CompressionMetadata metadata{CompressionType::INTEGER_BITPACKING, header.getDataByte()};
+        return metadata;
+    }
+
+    uint64_t compressNextPage(const uint8_t*& srcBuffer, uint64_t numValuesRemaining,
+        uint8_t* dstBuffer, uint64_t dstBufferSize,
+        const struct CompressionMetadata& metadata) const final;
+
+    void decompressFromPage(const uint8_t* srcBuffer, uint64_t srcOffset, uint8_t* dstBuffer,
+        uint64_t dstOffset, uint64_t numValues,
+        const struct CompressionMetadata& metadata) const final;
+
+protected:
+    // Read multiple values from within a chunk. Cannot span multiple chunks.
+    void getValues(const uint8_t* chunkStart, uint8_t pos, uint8_t* dst, uint8_t numValuesToRead,
+        const BitpackHeader& header) const;
+
+    inline const uint8_t* getChunkStart(
+        const uint8_t* buffer, uint64_t pos, uint8_t bitWidth) const {
+        return buffer + pos / CHUNK_SIZE * bitWidth;
+    }
+
+    common::LogicalType mLogicalType;
+};
+
+class BooleanBitpacking : public CompressionAlg {
+    static const common::LogicalType LOGICAL_TYPE;
+
+public:
+    BooleanBitpacking() = default;
+    BooleanBitpacking(const BooleanBitpacking&) = default;
+
+    void setValueFromUncompressed(uint8_t* srcBuffer, common::offset_t posInSrc, uint8_t* dstBuffer,
+        common::offset_t posInDst, const CompressionMetadata& metadata) const final;
+
+    void getValue(const uint8_t* buffer, common::offset_t posInBuffer, uint8_t* dst,
+        common::offset_t posInDst, const CompressionMetadata& metadata) const final;
+
+    static inline uint64_t numValues(uint64_t dataSize) { return dataSize * 8; }
+
+    inline CompressionMetadata startCompression(
+        const uint8_t* srcBuffer, uint64_t numValues) const override {
+        return CompressionMetadata{CompressionType::BOOLEAN_BITPACKING, 0};
+    }
+    uint64_t compressNextPage(const uint8_t*& srcBuffer, uint64_t numValuesRemaining,
+        uint8_t* dstBuffer, uint64_t dstBufferSize,
+        const struct CompressionMetadata& metadata) const final;
+
+    void decompressFromPage(const uint8_t* srcBuffer, uint64_t srcOffset, uint8_t* dstBuffer,
+        uint64_t dstOffset, uint64_t numValues, const CompressionMetadata& metadata) const final;
+};
+
+class CompressedFunctor {
+public:
+    CompressedFunctor(const CompressedFunctor&) = default;
+
+protected:
+    explicit CompressedFunctor(const common::LogicalType& logicalType) : copy{logicalType} {
+        switch (logicalType.getLogicalTypeID()) {
+        case common::LogicalTypeID::TIMESTAMP:
+        case common::LogicalTypeID::INT64: {
+            integerBitpacking = std::make_unique<IntegerBitpacking<int64_t>>();
+            break;
+        }
+        case common::LogicalTypeID::DATE:
+        case common::LogicalTypeID::INT32: {
+            integerBitpacking = std::make_unique<IntegerBitpacking<int32_t>>();
+            break;
+        }
+        default: {
```

Comment:
Let's guard the default case by throwing out `NotImplementedException` in advance.

Reply:
That's not the purpose of the switch though. This is just creating the 64/32-bit IntegerBitpacking class ahead of time in case it is used in `operator()` later, at which point we switch based on the compression type instead.
If anything, we need a way of asserting that any column using integer bitpacking has a type that is covered in this switch.

The IntegerBitpacking class no longer contains any data though, so I guess we could store the PhysicalTypeID instead and switch in `operator()` as the constructor should be more or less a noop. That would also remove the need for virtual dispatch.